indico/modules/events/editing/controllers/backend/timeline.py

```diff
@@ -162,7 +165,25 @@ def _process(self, action, comment):
             argmap['files'] = EditingFilesField(self.event, self.contrib, self.editable_type, allow_claimed_files=True,
                                                 required=True)
         args = parser.parse(argmap)
-        review_editable_revision(self.revision, session.user, action, comment, args['tags'], args.get('files'))
+        service_url = editing_settings.get(self.event, 'service_url')
+
+        revision = review_editable_revision(self.revision, session.user, action, comment, args['tags'],
+                                            args.get('files'))
+        publish = True
+        db.session.commit()
```

Comment:
I'm starting to dislike the fact that we need to commit because there may be new files uploaded during the review. This is a mess, but i guess we cannot avoid this.

Reply:
I agree. I was wondering that maybe we could discuss an alternative to this. Not sure how well it would work, but the idea would be to make this a pre-listener rather than a post-listener (before changes are committed). In sum, we would send the object to the service and only commit it after the service confirmation.

This comes with a couple of pitfalls like trying to modify an object that doesn't exist yet, but ideally, we could also receive everything as a response from the service (instead of doing new API calls from the service). Can't say if this is a better alternative because it is also less flexible, but it ensures nothing is committed in-between.

Reply:
The main reason why we commit first is so the microservice can retrieve files, right?

Could we maybe send URLs with a token that allows downloading the revision's files based on their UUID?
Basically, we'd add an endpoint like this (next to `RHFileInfo`):

```python
# /files/<uuid:uuid>/download?token=<token>
class RHFileDownload(RHFileBase):
    def _check_access(self):
        # this raises if the token is invalid/expired
        uuid = secure_serializer.loads(request.view_args['token'], salt='file-download', max_age=86400)
        if uuid != self.file.uuid:
            raise Forbidden
    
    def _process(self):
        return self.file.send()
```

and then build an url pointing to that endpoint and send this as the `external_url` when serializing the revision's files for the microservice.

For basic operations the microservice should perform immediately, we'd use returned data from the request to the microservice, and for anything else the microservice would be expected to perform then asynchronously to happen after the commit (we could add an endpoint to check if a given revision already exists, or the microservice could simply retry if necessary).

Reply:
Exactly, this would be a much cleaner alternative. There are other use-cases where we might need to do something immediately, such as adding a comment, maybe even tags? But we can always make that the response of the service and perform them immediately on the first commit. Allowing the service to modify any of the attributes of the object we're about to commit.

Reply:
I would say anything that must be done immediately or affect the action being performed itself (e.g. modify the set of tags during a review), should be done by returning data from the microservice. And anything beyond that should be done later, asynchronously.

Reply:
I think the only place where it certainly makes sense to keep the intermediate commit is when *enabling* the microservice initially.